propolis/src/vmm/hdl.rs

```diff
@@ -33,35 +32,40 @@ pub fn create_vm(name: impl AsRef<str>, force: bool) -> Result<VmmHdl> {
 
 #[cfg(target_os = "illumos")]
 fn create_vm_impl(name: &str, force: bool) -> Result<VmmHdl> {
+    if name.len() >= bhyve_api::VM_MAX_NAMELEN {
+        return Err(Error::from_raw_os_error(libc::EINVAL));
+    }
     let ctl = OpenOptions::new()
         .write(true)
         .custom_flags(libc::O_EXCL)
         .open(bhyve_api::VMM_CTL_PATH)?;
-    let namestr = CString::new(name)
-        .or_else(|_x| Err(Error::from_raw_os_error(libc::EINVAL)))?;
-    let nameptr = namestr.as_ptr();
     let ctlfd = ctl.as_raw_fd();
 
-    let res = unsafe { libc::ioctl(ctlfd, bhyve_api::VMM_CREATE_VM, nameptr) };
+    let mut req = bhyve_api::vm_create_req::default();
+    req.name[..name.len()].copy_from_slice(name.as_bytes());
+    let res = unsafe { libc::ioctl(ctlfd, bhyve_api::VMM_CREATE_VM, &req) };
```

Comment:
(nit - kinda up to you)

Here and below, we have name length checks before copying the name into the C struct.
I think it makes sense to keep those two things as tightly coupled as possible, which could be done by an `impl`-ed `fn` that both checks the length and copies from a slice. I think this would make it harder for callers to:

- Check against the wrong length (off by ones are super easy here)
- Forget checking the length at all

Reply:
Added some additional comment context on that.

Reply:
I've tightened it up a bit.  Returning the error directly to the client was belt-and-suspenders anyways.  I also considered a `new` fn on those two structs which took the name and did the appropriate copying, since it doesn't make sense to create a struct w/o a name, if that makes more sense